src/rest_framework_dso/embedding.py

```diff
@@ -0,0 +1,192 @@
+
+It's written in an observer/listener style, allowing the embedded data to be determined
+during the rendering. Instead of having to load the main results in memory for analysis,
+the main objects are inspected while they are consumed by the output stream.
+"""
+from typing import Dict, Optional, Union
+
+from rest_framework import serializers
+from rest_framework.exceptions import ParseError, PermissionDenied
+
+from dso_api.dynamic_api.permissions import fetch_scopes_for_model
+from rest_framework_dso.fields import AbstractEmbeddedField, EmbeddedField
+from rest_framework_dso.serializer_helpers import ReturnGenerator, peek_iterable
+
+EmbeddedFieldDict = Dict[str, AbstractEmbeddedField]
+
+
+def get_expanded_fields(  # noqa: C901
+    parent_serializer: serializers.Serializer, fields_to_expand: Union[list, bool]
+) -> Dict[str, EmbeddedField]:
+    """Find the expanded fields in a serializer that are requested.
+    This translates the ``_expand`` query into a dict of embedded fields.
+    """
+    if not fields_to_expand:
+        return {}
+
+    allowed_names = getattr(parent_serializer.Meta, "embedded_fields", [])
+    auth_checker = getattr(parent_serializer, "get_auth_checker", lambda: None)()
+    embedded_fields = {}
+
+    # ?_expand=true should expand all names
+    auto_expand_all = fields_to_expand is True
+    if auto_expand_all:
+        fields_to_expand = allowed_names
+
+    for field_name in fields_to_expand:
+        if field_name not in allowed_names:
+            raise _expand_parse_error(allowed_names, field_name) from None
+
+        # Get the field
+        try:
+            field = getattr(parent_serializer, field_name)
+        except AttributeError:
+            raise RuntimeError(
+                f"{parent_serializer.__class__.__name__}.{field_name}"
+                f" does not refer to an embedded field."
+            ) from None
+
+        # Check access via scopes (NOTE: this is higher-level functionality)
+        scopes = fetch_scopes_for_model(field.related_model)
+        if auth_checker and not auth_checker(*scopes.table):
+            # Not allowed, silently drop for _expand=true request.
+            if auto_expand_all:
+                continue
+
+            # Explicitly mentioned, raise error.
+            raise PermissionDenied(f"Eager loading not allowed for field '{field_name}'")
+
+        embedded_fields[field_name] = field
+
+    return embedded_fields
+
+
+def _expand_parse_error(allowed_names, field_name):
+    """Generate the proper exception for the invalid expand name"""
+    available = ", ".join(sorted(allowed_names))
+    if not available:
+        return ParseError("Eager loading is not supported for this endpoint")
+    else:
+        return ParseError(
+            f"Eager loading is not supported for field '{field_name}', "
+            f"available options are: {available}"
+        )
+
+
+class ObservableIterator:
```

Comment:
Brand new class. Hence could do with some type annotations. Clearly from the name it suggests its an `Iterable`, but I'm not sure yet what it iterates over. Something like this would help:
```
class ObservableIterator(Iterator[Fubar]):
```
Where `Fubar` is replaced by whatever it is `ObservableIterator` iterates over.

This also means that you should probably type annotate the member functions as well.

```python
def __iter__(self) -> Iterator[Fubar]:
   ...

def __next__(self) -> Fubar:
   ...
```

Reply:
Done!